### PR TITLE
Add initial support for Linux MD RAID

### PR DESCRIPTION
**Pull request by Schamper**
Not quite all features implemented yet (i.e. handling missing disks). Most other common features should work though.

(DIS-1388)

### CODE REVIEW

File: dissect/volume/md/stream.py [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+            r.append(device.fh.read(length))
+
+            num_sectors -= read_sectors
+            length -= read_length
+            offset_sector += read_sectors
+
+        return b"".join(r)
+
+
+def ffz(val: int) -> int:
```
Comment (pyrco) [inline dissect/volume/md/stream.py:393]: A bit of context on what this function is/does would be good I think.

File: dissect/volume/md/stream.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+from dissect.util.stream import AlignedStream, MappingStream
+
+from dissect.volume.exceptions import MDError
+from dissect.volume.md.c_md import c_md
+
+if TYPE_CHECKING:
+    from dissect.volume.md.md import MD, Device
+
+
+def create_linear_stream(md: MD) -> BinaryIO:
```
Comment (pyrco) [inline dissect/volume/md/stream.py:16]: Would it be possible to have this as a `MappingStream` subclass? That would make the API for all the types of MD arryas more consistent.

File: dissect/volume/md/md.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+from dissect.volume.md.c_md import c_md
+from dissect.volume.md.stream import (
+    RAID0Stream,
+    RAID10Stream,
+    RAID456Stream,
+    create_linear_stream,
+)
+
+
+class MD:
```
Comment (pyrco) [inline dissect/volume/md/md.py:22]: Can you add some generic class doc strings? So it is a bit more clear which class to use for what.

File: dissect/volume/md/stream.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+class RAID0Stream(AlignedStream):
+    def __init__(self, md: MD):
+        self.md = md
+        self.devices = {dev.raid_disk: dev for dev in self.md.devices if dev.raid_disk is not None}
+        if len(self.devices) != md.raid_disks:
+            raise MDError(f"Missing disks in RAID0 set {self.md.uuid} ({self.md.name})")
+
+        devices = sorted(self.devices.values(), key=operator.attrgetter("raid_disk"))
+        num_strip_zones = 0
+        for dev1 in devices:
```
Comment (pyrco) [inline dissect/volume/md/stream.py:48]: A bit more comments on the the flow of the different algorithms would help reviewing ;-)

**Review by pyrco — CHANGES_REQUESTED** [review]

File: dissect/volume/md/stream.py (hunk 4 above)
Comment (Schamper) [reply dissect/volume/md/stream.py:48]: There aren't any comments in the original source either so I barely understand how it works myself :smile:

File: dissect/volume/md/stream.py [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+
+    def __init__(self, md: MD):
+        self.md = md
+        self.devices = {dev.raid_disk: dev for dev in self.md.devices if dev.raid_disk is not None}
+        if len(self.devices) != md.raid_disks:
+            raise MDError(f"Missing disks in RAID0 set {self.md.uuid} ({self.md.name})")
+
+        # Determine how many strip zones we need to construct
+        # If a RAID0 set consists of devices with different sizes, additional strip zones
+        # may exist on the larger devices but not on the smaller ones
```
Comment (pyrco) [inline dissect/volume/md/stream.py:49]: If I understand correctly how `strip_zones` are defined, wouldn't a nicer algorithm be something along the lines of:

```
devices = sorted(self.devices.values(), key=operator.attrgetter("sectors"))
device_sizes = [device.sectors for device in devices]

next_device_size_offset = 0
next_device_size = device_sizes[next_device_size_offset]

zones = [Zone(next_device_size * len(devices), 0, devices)]

while True:
    prev_device_size_offset = next_device_size_offset
    prev_device_size = next_device_size

    next_device_size_offset = bisect.bisect(device_sizes, prev_device_size)

    if next_device_size_offset >= len(devices):
        break

    next_device_size = device_sizes[next_device_size_offset]

    strip_size = next_device_size - prev_device_size
    strip_devices = devices[next_device_size_offset:]

    zone_end = strip_size * len(strip_devices) + zones[-1].zone_end
    dev_start = ?? 
    zones.append(Zone(zone_end, dev_start, strip_devices))
```

**Review by pyrco — CHANGES_REQUESTED** [review]

File: dissect/volume/md/stream.py (hunk 5 above)
Comment (Schamper) [reply dissect/volume/md/stream.py:49]: I tried to mimick the original algorithm as much as possible. With your suggestion you lose the importance of the `raid_disk` ordering, which is the index a device has in the RAID set. It's my understanding that the algorithm doesn't try to create the _optimal_ strip zones, but it tries to create strip zones within the device ordering of the RAID set.

**Review by pyrco — CHANGES_REQUESTED** [review]

File: dissect/volume/md/stream.py [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+                num_strip_zones += 1
+
+        # Determine the smallest device and calculate the overal volume size at the same time
+        size = 0
+        smallest = None
+        for dev in devices:
+            if not smallest or rounded_sectors[dev] < rounded_sectors[smallest]:
+                smallest = dev
+
+            size += rounded_sectors[dev] & ~(self.md.chunk_sectors - 1)
```
[the remark this answers is not in the document]
Comment (Schamper) [reply dissect/volume/md/stream.py:81]: Added a comment.

**Comment by mischw** [comment]
That sounds like a great addition!
Since I recently came across [DDF](https://www.snia.org/tech_activities/standards/curr_standards/ddf), is that something you plan to support too in the future?

**Comment by Schamper** [comment]
> That sounds like a great addition! Since I recently came across [DDF](https://www.snia.org/tech_activities/standards/curr_standards/ddf), is that something you plan to support too in the future?

Almost done already :) just some small cleanup and creating some additional unit tests/fixing some edge cases left. IMSM (Intel RAID) is next. Also rewrote LVM2 (#20) to allow for more easily supporting the other modes in there besides the default striped one (and supporting thin-provisioned as a an initial test!), so expect better support there too.

**Comment by mischw** [comment]
> 



> > That sounds like a great addition! Since I recently came across [DDF](https://www.snia.org/tech_activities/standards/curr_standards/ddf), is that something you plan to support too in the future?
> 
> Almost done already :) just some small cleanup and creating some additional unit tests/fixing some edge cases left. IMSM (Intel RAID) is next. Also rewrote LVM2 (#20) to allow for more easily supporting the other modes in there besides the default striped one (and supporting thin-provisioned as an initial test!), so expect better support there too.

That is great news! Looking forward to test it :)

File: dissect/volume/md/md.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        if sb_major == 1:
+            self.sb = c_md.mdp_superblock_1(fh)
+        elif sb_major == 0:
+            self.sb = c_md.mdp_super_t(fh)
+        else:
+            raise ValueError(f"Invalid MD version at {sb_offset:#x}: {sb_major}.{sb_minor}")
+
+        if self.sb.major_version == 1:
+            self.set_uuid = UUID(bytes_le=self.sb.set_uuid)
+            self.set_name = self.sb.set_name.split(b"\x00", 1)[0].decode(errors="surrigateescape")
```
Comment (pyrco) [inline dissect/volume/md/md.py:100]:
```suggestion
            self.set_name = self.sb.set_name.split(b"\x00", 1)[0].decode(errors="surrogateescape")
```

File: dissect/volume/md/stream.py [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+
+        # Determine the smallest device and calculate the overal volume size at the same time
+        size = 0
+        smallest = None
+        for dev in devices:
+            if not smallest or rounded_sectors[dev] < rounded_sectors[smallest]:
+                smallest = dev
+
+            # Calculate a rounded up size of the RAID0 set in sectors
+            size += rounded_sectors[dev] & ~(self.md.chunk_sectors - 1)
```
Comment (pyrco) [inline dissect/volume/md/stream.py:81]: If I calculate correct, this is always rounded _down_ to the nearest `md.chunk_sectors`.

`rounded_sectors` is already rounded down to the nearest `device.chunk_sectors`. Say `device.chunk_sectors = 8`, for a number of different  `rounded_sectors` with an `md.chunck_sectors = 16`, this gives:

```python
8 & ~(16 - 1) == 0
16 & ~(16 - 1) == 16
24 & ~(16 - 1) == 16
32 & ~(16 - 1) == 32
```

File: dissect/volume/raid/raid.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            return LinearStream(self)
+        elif self.level == Level.RAID0:
+            return RAID0Stream(self)
+        elif self.level == Level.RAID1:
+            # Don't really care which mirror to read from, so just open the first disk
+            return self.physical_disks[0][1].open()
+        elif self.level in (Level.RAID4, Level.RAID5, Level.RAID6):
+            return RAID456Stream(self)
+        elif self.level == Level.RAID10:
+            return RAID10Stream(self)
```
Comment (pyrco) [inline dissect/volume/raid/raid.py:71]:
```suggestion
            return RAID10Stream(self)
        else:
           raise ?
```

File: dissect/volume/md/stream.py [hunk 10]
```diff
@@ -0,0 +1,429 @@
+from __future__ import annotations
```
Comment (pyrco) [inline dissect/volume/md/stream.py:1]: Seems like this file's functionality is (almost) completely superseded by `raid/stream.py`?

**Review by pyrco — CHANGES_REQUESTED** [review]

File: dissect/volume/md/stream.py (hunk 10 above)
Comment (Schamper) [reply dissect/volume/md/stream.py:1]: Oops yeah, this file was supposed to have been deleted.

File: dissect/volume/md/stream.py (hunk 8 above)
Comment (Schamper) [reply dissect/volume/md/stream.py]: I don't think this is relevant anymore as this file has been superseded by `raid/stream.py` where this was removed.

File: dissect/volume/raid/stream.py [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+                    offset = offset - self.zones[i - 1].zone_end
+                return zone, offset
+        return None, None
+
+    def _read(self, offset: int, length: int) -> bytes:
+        result = []
+
+        stripe_size = self.virtual_disk.stripe_size
+        # offset_sector = offset // SECTOR_SIZE
+        # num_sectors = (length + SECTOR_SIZE - 1) // SECTOR_SIZE
```
Comment (pyrco) [inline dissect/volume/raid/stream.py:160]:
```suggestion
```

**Review by pyrco — CHANGES_REQUESTED** [review]

**Review by pyrco — APPROVED** [review]
